File: src/org/cerion/symcalc/expression/function/integer/EvenQ.java
package org.cerion.symcalc.expression.function.integer;

import org.cerion.symcalc.exception.ValidationException;
import org.cerion.symcalc.expression.BoolExpr;
import org.cerion.symcalc.expression.Expr;
import org.cerion.symcalc.expression.FunctionExpr;
import org.cerion.symcalc.expression.number.IntegerNum;

public class EvenQ extends FunctionExpr {

    public EvenQ(Expr ...e) {
        super(FunctionType.EVENQ, e);
    }

    @Override
    protected Expr evaluate() {

        if(!get(0).isInteger())
            return BoolExpr.FALSE;

        IntegerNum N = (IntegerNum)get(0);
        if(N.IsEven())
            return BoolExpr.TRUE;

        return BoolExpr.FALSE;
    }

    @Override
    public void validate() throws ValidationException {
        validateParameterCount(1);
    }
}

File: src/org/cerion/symcalc/expression/function/integer/EvenQ.kt
package org.cerion.symcalc.expression.function.integer

import org.cerion.symcalc.exception.ValidationException
import org.cerion.symcalc.expression.BoolExpr
import org.cerion.symcalc.expression.Expr
import org.cerion.symcalc.expression.FunctionExpr

class EvenQ(vararg e: Expr) : FunctionExpr(FunctionExpr.FunctionType.EVENQ, *e) {

    override fun evaluate(): Expr {
        if (!get(0).isInteger)
            return BoolExpr.FALSE

        return if (get(0).asInteger().isEven) BoolExpr.TRUE else BoolExpr.FALSE
    }

    @Throws(ValidationException::class)
    override fun validate() {
        validateParameterCount(1)
    }
}
